homeassistant/components/lgthinq/__init__.py

```diff
@@ -0,0 +1,129 @@
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import CONF_CONNECT_CLIENT_ID, DEFAULT_COUNTRY
+from .coordinator import DeviceDataUpdateCoordinator, async_setup_device_coordinator
+
+type ThinqConfigEntry = ConfigEntry[dict[str, DeviceDataUpdateCoordinator]]
+
+PLATFORMS = [Platform.SWITCH]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ThinqConfigEntry) -> bool:
+    """Set up an entry."""
+    access_token = entry.data.get(CONF_ACCESS_TOKEN)
+    client_id = entry.data.get(CONF_CONNECT_CLIENT_ID)
+    country_code = entry.data.get(CONF_COUNTRY, DEFAULT_COUNTRY)
+
+    thinq_api = ThinQApi(
+        session=async_get_clientsession(hass),
+        access_token=access_token,
+        country_code=country_code,
+        client_id=client_id,
+    )
+
+    # Setup coordinators and register devices.
+    await async_setup_coordinators(hass, entry, thinq_api)
+
+    # Set up all platforms for this device/entry.
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    # Clean up devices they are no longer in use.
+    async_cleanup_device_registry(hass, entry)
+
+    return True
+
+
+async def async_setup_coordinators(
+    hass: HomeAssistant,
+    entry: ThinqConfigEntry,
+    thinq_api: ThinQApi,
+) -> None:
+    """Set up coordinators and register devices."""
+    entry.runtime_data = {}
+
+    # Get a device list from the server.
+    try:
+        device_list = await thinq_api.async_get_device_list()
+    except ThinQAPIException as exc:
+        raise ConfigEntryError(exc.message) from exc
+
+    if not device_list or not isinstance(device_list, Collection):
+        return
+
+    # Setup coordinator per device.
+    coordinator_list: list[DeviceDataUpdateCoordinator] = []
+    task_list = [
+        hass.async_create_task(async_setup_device_coordinator(hass, thinq_api, device))
+        for device in device_list
+    ]
+    if task_list:
+        task_result = await asyncio.gather(*task_list)
+        for coordinators in task_result:
+            if coordinators:
+                coordinator_list += coordinators
+
+    # Register devices.
+    async_register_devices(hass, entry, coordinator_list)
+
+
+@callback
+def async_register_devices(
+    hass: HomeAssistant,
+    entry: ThinqConfigEntry,
+    coordinator_list: list[DeviceDataUpdateCoordinator],
+) -> None:
+    """Register devices to the device registry."""
+    device_registry = dr.async_get(hass)
+
+    for coordinator in coordinator_list:
+        device_entry = device_registry.async_get_or_create(
+            config_entry_id=entry.entry_id,
+            **coordinator.device_info,
+        )
+        _LOGGER.debug(
+            "Register device: device_id=%s, device_entry_id=%s",
+            coordinator.device_api.device_id,
+            device_entry.id,
+        )
+        entry.runtime_data[device_entry.id] = coordinator
```

Comment:
When adding entities with a `device_info`, the device is automatically created. So we only need to do manual device creation when we either don't have the full info in the entity, or the device doesn't have entities (for example a hub). I don't think either are the case, so should we remove this?